_posts/2015-09-30-september-newsletter.md

```diff
@@ -0,0 +1,274 @@
+
+- Structured standardized file format based on msgpack, allowing easier creation
+  of plugins/tools that perform arbitrary manipulation in any language that
+  supports msgpack.
+- Binary serialization format allows efficient/safe storage of arbitrary data.
+- Multiple Neovim instances will be able to share data in real time:
+  - Command history
+  - Register contents. eg: yank in one instance and paste in another. This is
+    already possible with the clipboard register, but this may not be available
+    in certain environments.
+
+While the [PR][shada-pr] is very big, @ZyX-I has taken care of adding great test
+coverage. Great work @ZyX-I!
+
+
+### Wrapping the event loop layer
+
+It's no secret that libuv is the event loop library used by Neovim, and it is
+what makes it possible for us to implement features that require asynchronous
+communication(not initiated by the user) with the editor with ease.
+Unfortunately due to how Neovim code is currently organized, integrating the
+existing code with libuv was not a trivial task.
+
+The basic idea is that Neovim receives arbitrary events when it is polling for
+user input, but these events can't be processed immediately because Neovim can
+be in a state that simply can't handle arbitrary actions. So if Neovim receives
+an event while checking user input, it will put the event in a queue for later
+processing.
+
+One example is illustrated in the following scenario: Neovim checks if the user
+typed ctrl+c while the regexp engine is executing, but it can't process the
+event as it is received because it may want to execute vimscript that calls the
+regexp engine again, and the engine is not reentrant since it relies heavily on
+global variables. So it has to postpone the event for when it's safe, and
+determining when it's safe to process events is itself another problem.
+
+Another complication of integrating with libuv is that sometimes Neovim must
+only process events from a certain source. For example, when Neovim sends a
+msgpack-rpc call, it should only process events that come from:
+
+- the user(eg: ctrl+c to interrupt the call)
+- the file descriptor that received the msgpack-rpc call(which can be from a child
+  process stdio from a socket)
+
+To allow this kind of selective event processing, Neovim must maintain multiple
+queues that integrate with each other, and the logic to do this is very
+repetitive. In one of my [latest PRs][event-loop-pr], some libuv "classes" were
+wrapped in a way that makes managing these queues much easier.
+
+### Jemalloc
+
+[jemalloc][jemalloc], a high performance general purpose memory allocator, is
+now used by default. Since Neovim makes heavy use of dynamic queues(see above),
+it relies on `malloc(3)` a lot more than Vim, so it is important to use a fast
+implementation that has consistent performance across platforms.
+
+In a recent [pr][jemalloc-4-pr], @fwalch modified upgraded the jemalloc version
```

Comment:
That was me, not @fwalch ;) He concurrently fixed an issue we had with caching dependencies in the build system, perhaps that can be mentioned aswell.